cpp/src/arrow/util/uri.cc

```diff
@@ -208,6 +208,8 @@ std::string Uri::path() const {
   return std::move(ss).str();
 }
 
+std::string Uri::extension() const { return impl_->path_segments_.back().to_string(); }
```

Comment:
`path_segments_` could be empty, looking at the implementation.

Reply:
Also, is the last path segment really the extension? I would've expected it to be the filename…